### ISSUE TITLE
Debian family installer sets up an account with a hardcoded password in the default config

### ISSUE THREAD

**Issue by danderson** [digitalrebar/provision#872]
Looking through net-seed.tmpl, there is the following:

```
        d-i passwd/make-user boolean true
        d-i passwd/user-uid string {{if .ParamExists "provisioner-default-uid"}}{{.Param "provisioner-default-uid"}}{{else}}1000{{end}}
        d-i passwd/user-fullname string {{if .ParamExists "provisioner-default-fullname"}}{{.Param "provisioner-default-fullname"}}{{else if .ParamExists "provisioner-default-user"}}{{.Param "provisioner-default-user"}}{{else}}Rocket Skates{{end}}
        d-i passwd/username string {{if .ParamExists "provisioner-default-user"}}{{.Param "provisioner-default-user"}}{{else}}rocketskates{{end}}
        d-i passwd/user-password-crypted password {{if .ParamExists "provisioner-default-password-hash"}}{{.Param "provisioner-default-password-hash"}}{{else}}$6$drprocksdrprocks$upAIK9ynEEdFmaxJ5j0QRvwmIu2ruJa1A1XB7GZjrnYYXXyNr4qF9FttxMda2j.cmh.TSiLgn4B/7z0iSHkDC1{{end}}
        d-i user-setup/allow-password-weak boolean true
```

In DRP's default setup, this means that every debian or ubuntu machine I install will be configured with a "rocketskates" user, using a password shared by all DRP users.

This is a very poor default. A better default would be to set passwd/make-user=false if the provisioner params aren't present. That would better match the user intent of "if I didn't ask you for a user account, don't give me a user account."

I'm willing to send a patch for this, if you believe it's a reasonable approach.

**Comment by danderson** [comment]
Turns out the Centos kickstart config does the same thing.

And yes, it's documented at http://provision.readthedocs.io/en/latest/doc/configuring.html#default-template-identity , but I feel like it really should have a big red flashing banner that says "OS installs are unsafe by default" :/

So I guess this is actually a request to not create any user account unless explicitly requested by the user. WDYT?

**Comment by galthaus** [comment]
Well, it isn't really hard-coded.  It is parameterized with a default.  The pattern is pretty common in the templates.  This is really tough because we are trying to have a balance between knowledgable security-aware operators that will setup default ssh keys, turn off remote access, and set a root/master account password with console only access and the people coming into the field that are just trying to learn about provisioning in a easy to try-out non-production environment.  For now, we've picked defaults that a balance.  Ideally, the root account doesn't have net access (unless you change another parameter) but has a password that can be set globally or on a machine-by-machine basis from the parameters.  We are trying to get people into the basics of provisioning and them onto security practices and reuse, rebuild, immutable image deploys and all those other things.

**Comment by galthaus** [comment]
I think the problem with the passwd/make-user=false change would be that we would now expect users to inject ssh keys with a parameter.  Not that hard, but an additional hurdle for the newbie.

I do think we need an option that sets passwd/make-user=false in the seed for those that really don't want that user, but have ssh keys in place to continue operations.

**Comment by danderson** [comment]
Definitely understand the desire to balance perfect defaults and ease of getting started. I guess my main concern is that, after following the quickstart and kicking the tires for several hours, I still wasn't aware that this default existed - it's nowhere in the UX or APIs, because that default is hardwired in a template file at the bottom of the stack.

So, even though I'd been walking all over the API and feature surface, it's only when I got down into dissecting the preseed that I found it. Once I did, I found the note in the manual that points this out, but I came away from the experience thinking "gee, I wish DRP had made more noise about this."

Here's a suggestion for an experience that would have surprised me less: instead of hardcoding the default in the kickseed templates, place those defaults as params in the "global" profile when DRP first populates its empty datadir. Then, add an item in the "System Wizard" UX that says "hey, just so you know, here's the default username/password we provision on machine installs" with a link to the global profile UX where it can be easily changed (or even build a custom UX for the initial "give me a username/password and I'll update the global profile for you" wizard flow, but that's more work :) ).

I think this would improve things in two major ways:

- When I'm kicking the tires, I'm likely to be using the UX rather than dive straight into drpcli and the API. The UX drops me straight onto the wizard checklist, so the information is front and center - if I'm just exploring, it tells me right away what login to use, and if I don't like that, it takes me straight to the right place to change it to a safe default.
- Even if I somehow miss that, at some point I'm going to start playing with profiles to customize my machines, and I'll organically come across those entries in the global profile, go "huh, what's that about?", and not have to dig all the way down to the templates to figure it out.


Separately, agreed that for getting started, passwd/make-user=false isn't friendly. But, I also agree that adding a param for that would be excellent for my particular use case :)

**Comment by galthaus** [comment]
From Twitter, Dave and I discussed this some more.  The net of it was a suggestion to have ubuntu not create a user by default, set the root password and keys, document that we are normalizing that the default config path, and then potentially add some UX wizard helpers to indicate things like root password defaults and ssh keys not be set and how to set them.

**Comment by VictorLowther** [comment]
hrm, not to go too far down that raboot hole, but we go thorugh the trouble of not setting a root password by default in Ubuntu because they like defaulting to "you must sudo to do anything as root", and we wanted to give users something that vaguely approximates what would happen by default in the OS install path.

**Comment by danderson** [comment]
Yeah, it comes down to whether it's better to match upstream's intent by default, or provide DRP-specific defaults that make sense for managed infrastructure. I personally lean towards the latter, since preseed comes with all the right knobs to enact that even if the GUI installer pushes you towards users and sudo. But both options have merit.

**Comment by zehicle** [comment]
There's a challenge about if you have unsafe defaults that help new users.  Our current assumption is that production users will learn enough to make safe changes.

We'll continue to evaluate; however, at this time I  believe we want to keep initial use easier.

**Comment by galthaus** [comment]
I'm going to leave this open as a marker issue.
